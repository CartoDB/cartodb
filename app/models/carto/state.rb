require 'json'
require_relative './carto_json_serializer'

class Carto::State < ActiveRecord::Base
  belongs_to :visualization, class_name: Carto::Visualization

  serialize :json, ::Carto::CartoJsonSymbolizerSerializer
  validates :json, carto_json_symbolizer: true

  after_initialize :ensure_json

  def self.columns
    super.reject { |c| c.name == 'user_id' }
  end

  private

  def ensure_json
    self.json ||= Hash.new
  end
end
