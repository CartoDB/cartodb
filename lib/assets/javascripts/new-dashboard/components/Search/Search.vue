<template>
  <form class="navbar-search" :class="{'is-search-open': isSearchOpen}" autocomplete="off" @submit.stop.prevent="onFormSubmit">
    <input type="text"
           v-model.trim="searchTerm"
           ref="searchInput"
           class="title is-small is-regular navbar-searchInput"
           :class="{ 'navbar-searchInput--filled': isFilled }"
           :placeholder="placeholder"
           @focus="onInputFocus"
           @blur="onInputBlur">
    <SearchSuggestions :query="searchTerm" :isOpen="isInputFocused && isFilled" @pageChange="resetInput"/>
  </form>
</template>

<script>
import SearchSuggestions from './Suggestions/SearchSuggestions';

export default {
  name: 'Search',
  components: {
    SearchSuggestions
  },
  data () {
    return {
      searchTerm: '',
      isInputFocused: false
    };
  },
  props: {
    isSearchOpen: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    placeholder () {
      if (this.isInputFocused) {
        return this.$t('SearchComponent.placeholder.active');
      }

      return this.$t('SearchComponent.placeholder.default');
    },
    isFilled () {
      return Boolean(this.searchTerm);
    }
  },
  methods: {
    onInputFocus () {
      this.isInputFocused = true;
    },

    onInputBlur () {
      this.isInputFocused = false;
    },

    onFormSubmit () {
      this.blurInput();

      if (this.searchTerm.includes(':')) {
        this.$router.push({ name: 'tagSearch', params: { tag: this.searchTerm.substring(1) } });
      } else if (this.searchTerm) {
        this.$router.push({ name: 'search', params: { query: this.searchTerm } });
      }

      this.searchTerm = '';
    },

    blurInput () {
      this.$refs.searchInput.blur();
    },

    resetInput () {
      this.searchTerm = '';
      this.blurInput();
    }
  }
};
</script>

<style lang="scss" scoped>
@import 'stylesheets/new-dashboard/variables';

.navbar-search {
  @media (max-width: $layout-mobile) {
    visibility: hidden;
    opacity: 0;
  }

  &::after {
    content: '';
    position: absolute;
    top: 50%;
    left: 16px;
    width: 18px;
    height: 18px;
    transform: translate3d(0, -50%, 0);
    background-image: url("../../assets/icons/navbar/loupe.svg");
    background-repeat: no-repeat;
    background-position: center;
  }

  &:hover {
    &::after {
      background-image: url("../../assets/icons/navbar/loupe_primary.svg");
    }

    .navbar-searchInput {
      &::placeholder {
        color: $primary-color;
      }

      &:focus {
        &::placeholder {
          color: $text-secondary-color;
        }
      }
    }
  }

  &:focus-within {
    &::after {
      background-image: url("../../assets/icons/navbar/loupe.svg");
    }
  }
}

.navbar-searchInput {
  width: 120px;
  height: 36px;
  padding: 0 14px 0 42px;
  transition: width 0.3s cubic-bezier(0.4, 0.01, 0.165, 0.99);
  border: 0;
  border-radius: 18px;
  background-color: #FFF;

  &::placeholder {
    color: $text-secondary-color;
  }

  &:focus,
  &.navbar-searchInput--filled {
    width: 240px;
    outline: none;
  }
}

.navbar.is-search-open {
  .navbar-search {
    visibility: visible;
    position: absolute;
    left: 0;
    width: calc(100% - 66px);
    margin-left: 16px;
    opacity: 1;

    &::after {
      display: none;
    }
  }

  .navbar-searchInput {
    width: 100%;
    padding: 16px;
    border-radius: 3px;
    background-image: none;
  }
}
</style>
