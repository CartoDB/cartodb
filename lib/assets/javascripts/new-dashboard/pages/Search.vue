<template>
  <section class="page">
    <StickySubheader :is-visible="true" class="page-subheader">
      <span class="title" v-if="isFirstFetch">
        {{ $t('SearchPage.title.allFetching', { query: searchTerm || tag }) }}
        <span class="loading">
          <img svg-inline src="../assets/icons/common/loading.svg" class="loading__svg"/>
        </span>
      </span>
      <span class="title" v-else-if="searchTerm">{{ $tc('SearchPage.title.searchTerm', totalResults, { query: searchTerm }) }}</span>
      <span class="title" v-else-if="tag">{{ $tc('SearchPage.title.tag', totalResults, { query: tag }) }}</span>
    </StickySubheader>

    <div class="container grid">
      <div class="full-width">
        <section class="page-section" :class="{ 'has-pagination': hasMaps && mapsNumPages > 1 }" ref="maps">
          <div class="section-title grid-cell title is-medium">{{ $t('SearchPage.sections.maps') }}</div>

          <ul class="grid" v-if="isFetchingMaps">
            <li v-for="n in 3" :key="n" class="grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile map-element">
              <MapCardFake></MapCardFake>
            </li>
          </ul>

          <ul class="grid" v-if="!isFetchingMaps">
            <li v-for="map in maps" :key="map.id" class="grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile map-element">
              <MapCard :map=map :canHover=false></MapCard>
            </li>

            <div class="grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile is-caption text maps--empty" v-if="!hasMaps">
              {{ $t('SearchPage.emptyText.maps') }}
            </div>
          </ul>

          <Pagination
            v-if="hasMaps && mapsNumPages > 1"
            :page=mapsPage
            :numPages=mapsNumPages
            @pageChange="page => onPageChange('maps', page)"></Pagination>
        </section>

        <section class="page__section" ref="datasets">
          <div class="section-title grid-cell title is-medium">{{ $t('SearchPage.sections.data') }}</div>

          <ul class="grid-cell grid-cell--col12" v-if="!isFetchingDatasets">
            <li v-for="dataset in datasets" :key="dataset.id" class="dataset-item">
              <DatasetCard :dataset=dataset :canHover=false></DatasetCard>
            </li>

            <div class="is-caption text" v-if="!hasDatasets">
              {{ $t('SearchPage.emptyText.datasets') }}
            </div>
          </ul>

          <ul class="grid-cell grid-cell--col12" v-if="isFetchingDatasets">
            <li v-for="n in 3" :key="n" class="dataset-item">
              <DatasetCardFake></DatasetCardFake>
            </li>
          </ul>

          <Pagination
            class="pagination-element"
            v-if="hasDatasets && datasetsNumPages > 1"
            :page=datasetsPage
            :numPages=datasetsNumPages
            @pageChange="page => onPageChange('datasets', page)"></Pagination>
        </section>
      </div>
    </div>

  </section>
</template>

<script>
import StickySubheader from 'new-dashboard/components/StickySubheader';
import MapCard from 'new-dashboard/components/MapCard';
import MapCardFake from 'new-dashboard/components/MapCardFake';
import DatasetCard from 'new-dashboard/components/Dataset/DatasetCard';
import DatasetCardFake from 'new-dashboard/components/Dataset/DatasetCardFake';
import Pagination from 'new-dashboard/components/Pagination';
import updateSearchParams from 'new-dashboard/router/hooks/update-search-params';
import { mapState } from 'vuex';

const TWO_HEADERS_HEIGHT = 128;

export default {
  name: 'SearchPage',
  components: {
    DatasetCard,
    DatasetCardFake,
    MapCard,
    MapCardFake,
    Pagination,
    StickySubheader
  },
  beforeRouteUpdate (to, from, next) {
    this.$store.dispatch('search/resetState');
    this.isFirstFetch = true;

    updateSearchParams(to, from, next);
  },
  beforeRouteLeave (to, from, next) {
    this.$store.dispatch('search/resetState');
    next();
  },
  data () {
    return {
      isFirstFetch: true
    };
  },
  computed: {
    ...mapState({
      searchTerm: state => state.search.searchTerm,
      tag: state => state.search.tag,
      maps: state => state.search.maps.results,
      mapsPage: state => state.search.maps.page,
      mapsNumPages: state => state.search.maps.numPages,
      isFetchingMaps: state => state.search.maps.isFetching,
      datasets: state => state.search.datasets.results,
      datasetsPage: state => state.search.datasets.page,
      datasetsNumPages: state => state.search.datasets.numPages,
      isFetchingDatasets: state => state.search.datasets.isFetching,
      totalResults: state => state.search.maps.numResults + state.search.datasets.numResults
    }),
    hasMaps () {
      return Object.keys(this.maps || {}).length;
    },
    hasDatasets () {
      return Object.keys(this.datasets || {}).length;
    },
    allSectionsFetching () {
      return this.isFetchingMaps || this.isFetchingDatasets;
    }
  },
  methods: {
    onPageChange (section, page) {
      this.$store.dispatch('search/changeSectionPage', { section, page });
      this.scrollToSection(section);
    },
    isOutOfViewport (boundingClientRect) {
      return boundingClientRect.top < TWO_HEADERS_HEIGHT;
    },
    scrollToSection (section) {
      const sectionBoundingClientRect = this.$refs[section].getBoundingClientRect();
      const offsetDistance = TWO_HEADERS_HEIGHT + 16; // Two headers + margin

      if (!this.isOutOfViewport(sectionBoundingClientRect)) {
        return;
      }

      window.scrollBy({ top: sectionBoundingClientRect.top - offsetDistance, behavior: 'smooth' });
    }
  },
  watch: {
    allSectionsFetching (newValue) {
      if (newValue === false) {
        this.isFirstFetch = false;
      }
    }
  }
};
</script>

<style lang="scss" scoped>
@import 'stylesheets/new-dashboard/variables';

.page {
  padding-top: 192px;
}

.page-section {
  margin-bottom: 12px;

  &.has-pagination,
  &:last-child {
    margin-bottom: 48px;
  }
}

.page-subheader {
  background-color: $softblue;
}

.loading {
  margin-left: 24px;

  &__svg {
    width: 18px;
    vertical-align: text-top;

    g {
      stroke-width: 2px;
    }
  }
}

.full-width {
  width: 100%;
}

.section-title {
  margin-bottom: 16px;
}

.map-element {
  margin-bottom: 36px;
}

.pagination-element {
  margin-top: 36px;
}

.dataset-item {
  &:not(:last-child) {
    border-bottom: 1px solid $light-grey;
  }
}

.maps--empty {
  margin-bottom: 128px;
}
</style>
