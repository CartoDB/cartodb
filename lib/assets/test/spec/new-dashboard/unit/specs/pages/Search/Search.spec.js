import { shallowMount } from '@vue/test-utils';
import Search from 'new-dashboard/pages/Search';
import { store, fetchingStore, resultsStore } from './state';

describe('Search.vue', () => {
  describe('Rendering', () => {
    it('should render fetching state correctly', () => {
      const searchComponent = createSearchComponent({
        mocks: {
          $store: fetchingStore
        }
      });
      expect(searchComponent).toMatchSnapshot();
    });

    it('should render empty state correctly', () => {
      const searchComponent = createSearchComponent({
        data () {
          return {
            isFirstFetch: false
          };
        }
      });

      expect(searchComponent).toMatchSnapshot();
    });

    it('should render search results correctly', () => {
      const searchComponent = createSearchComponent({
        data () {
          return {
            isFirstFetch: false
          };
        },
        mocks: {
          $store: resultsStore
        }
      });

      expect(searchComponent).toMatchSnapshot();
    });
  });

  describe('Methods', () => {
    describe('onPageChange', () => {
      it('should dispatch changeSectionPage action and call scrollToSection', () => {
        const storeDispatchFn = jest.fn();
        const searchComponent = createSearchComponent({
          mocks: {
            $store: {
              dispatch: storeDispatchFn,
              ...store
            }
          }
        });

        spyOn(searchComponent.vm, 'scrollToSection');

        const section = 'maps';
        const page = 2;

        searchComponent.vm.onPageChange(section, page);

        expect(storeDispatchFn).toHaveBeenCalledWith('search/changeSectionPage', { section, page });
        expect(searchComponent.vm.scrollToSection).toHaveBeenCalledWith(section);
      });
    });

    describe('isOutOfViewport', () => {
      it("should return true if element's top position is less than headers offset", () => {
        const elementBoundingClientRect = {
          top: 119
        };

        const searchComponent = createSearchComponent();
        expect(searchComponent.vm.isOutOfViewport(elementBoundingClientRect)).toBe(true);
      });
    });

    describe('isOutOfViewport', () => {
      it("should return true if element's top position is less than headers offset", () => {
        // Note: Default boundingClientRect's top is 0, which is handy for us
        // given that our two headers size is 128
        const elementBoundingClientRect = {
          top: 127
        };

        const searchComponent = createSearchComponent();
        expect(searchComponent.vm.isOutOfViewport(elementBoundingClientRect)).toBe(true);
      });
    });

    describe('scrollToSection', () => {
      it('should call scrollBy if element is out of viewport', () => {
        // Note: Default boundingClientRect's top is 0, which is handy for us
        // given that our two headers size is 128
        const scrollBySpy = spyOn(window, 'scrollBy');
        const searchComponent = createSearchComponent();

        searchComponent.vm.scrollToSection('maps');

        expect(scrollBySpy).toHaveBeenCalledWith({ behavior: 'smooth', top: -144 });
      });

      it('should call scrollBy if element is inside the viewport', () => {
        const scrollBySpy = spyOn(window, 'scrollBy');
        const searchComponent = createSearchComponent();
        spyOn(searchComponent.vm, 'isOutOfViewport').and.returnValue(false);

        searchComponent.vm.scrollToSection('maps');

        expect(scrollBySpy).not.toHaveBeenCalled();
      });
    });
  });

  describe('Watch', () => {
    describe('allSectionsFetching', () => {
      it('should set isFirstFetch to false when allSectionsFetching changes to false', done => {
        const mockStore = { ...store };
        const searchComponent = createSearchComponent({
          data () {
            return {
              allSectionsFetching: true
            };
          },
          mocks: { $store: mockStore }
        });

        searchComponent.vm.allSectionsFetching = false;

        searchComponent.vm.$nextTick(() => {
          expect(searchComponent.vm.isFirstFetch).toBe(false);
          done();
        });
      });
    });
  });

  describe('Router Hooks', () => {
    describe('beforeRouteUpdate', () => {
      it('should dispatch resetState, and set isFirstFetch to true', () => {
        const storeDispatchFn = jest.fn();
        const $store = {
          dispatch: storeDispatchFn
        };

        const to = {
          params: {
            query: 'Query text'
          }
        };

        const updateSearchParams = () => {};

        Search.beforeRouteUpdate.apply({ $store, updateSearchParams }, [to, null, () => {}]);

        expect(storeDispatchFn).toHaveBeenCalledWith('search/resetState');
      });
    });
  });
});

function createSearchComponent (options = { mocks: {} }) {
  return shallowMount(Search, {
    computed: {
      searchTerm () {
        return 'Query text';
      },
      isNotificationVisible () {
        return false;
      }
    },
    ...options,
    mocks: {
      $t: key => key,
      $tc: key => key,
      $store: store,
      ...options.mocks
    }
  });
}
