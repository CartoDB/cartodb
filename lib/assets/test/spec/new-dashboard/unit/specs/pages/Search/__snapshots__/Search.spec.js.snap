// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search.vue Rendering should render empty state correctly 1`] = `
<section class="page">
  <stickysubheader-stub isvisible="true" class="page-subheader"><span class="title">SearchPage.title.searchTerm</span></stickysubheader-stub>
  <div class="container grid">
    <div class="full-width">
      <section class="page-section">
        <div class="section-title grid-cell title is-medium">SearchPage.sections.maps</div>
        <!---->
        <ul class="grid">
          <div class="grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile is-caption text maps--empty">
            SearchPage.emptyText.maps
          </div>
        </ul>
        <!---->
      </section>
      <section class="page__section">
        <div class="section-title grid-cell title is-medium">SearchPage.sections.data</div>
        <ul class="grid-cell grid-cell--col12">
          <div class="is-caption text">
            SearchPage.emptyText.datasets
          </div>
        </ul>
        <!---->
        <!---->
      </section>
    </div>
  </div>
</section>
`;

exports[`Search.vue Rendering should render fetching state correctly 1`] = `
<section class="page">
  <stickysubheader-stub isvisible="true" class="page-subheader"><span class="title">
      SearchPage.title.allFetching
      <span class="loading"><img svg-inline="" src="../assets/icons/common/loading.svg" class="loading__svg"></span></span></stickysubheader-stub>
  <div class="container grid">
    <div class="full-width">
      <section class="page-section">
        <div class="section-title grid-cell title is-medium">SearchPage.sections.maps</div>
        <ul class="grid">
          <li class="grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile map-element">
            <mapcardfake-stub></mapcardfake-stub>
          </li>
          <li class="grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile map-element">
            <mapcardfake-stub></mapcardfake-stub>
          </li>
          <li class="grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile map-element">
            <mapcardfake-stub></mapcardfake-stub>
          </li>
        </ul>
        <!---->
        <!---->
      </section>
      <section class="page__section">
        <div class="section-title grid-cell title is-medium">SearchPage.sections.data</div>
        <!---->
        <ul class="grid-cell grid-cell--col12">
          <li class="dataset-item">
            <datasetcardfake-stub></datasetcardfake-stub>
          </li>
          <li class="dataset-item">
            <datasetcardfake-stub></datasetcardfake-stub>
          </li>
          <li class="dataset-item">
            <datasetcardfake-stub></datasetcardfake-stub>
          </li>
        </ul>
        <!---->
      </section>
    </div>
  </div>
</section>
`;

exports[`Search.vue Rendering should render search results correctly 1`] = `
<section class="page">
  <stickysubheader-stub isvisible="true" class="page-subheader"><span class="title">SearchPage.title.searchTerm</span></stickysubheader-stub>
  <div class="container grid">
    <div class="full-width">
      <section class="page-section has-pagination">
        <div class="section-title grid-cell title is-medium">SearchPage.sections.maps</div>
        <!---->
        <ul class="grid">
          <li class="grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile map-element">
            <mapcard-stub map="[object Object]"></mapcard-stub>
          </li>
          <li class="grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile map-element">
            <mapcard-stub map="[object Object]"></mapcard-stub>
          </li>
          <!---->
        </ul>
        <pagination-stub page="1" numpages="2"></pagination-stub>
      </section>
      <section class="page__section">
        <div class="section-title grid-cell title is-medium">SearchPage.sections.data</div>
        <ul class="grid-cell grid-cell--col12">
          <li class="dataset-item">
            <datasetcard-stub dataset="[object Object]"></datasetcard-stub>
          </li>
          <li class="dataset-item">
            <datasetcard-stub dataset="[object Object]"></datasetcard-stub>
          </li>
          <!---->
        </ul>
        <!---->
        <pagination-stub page="1" numpages="2" class="pagination-element"></pagination-stub>
      </section>
    </div>
  </div>
</section>
`;
